### PR TITLE
Addrdb code deduplication + Native support for FLATDATA + Add file syncing logging and error handling

### PR DESCRIPTION
**Pull request by furszy**
Digging down the peers and ban databases commits path in upstream found some good stuff, back ported the following PRs:

* #9216.
* #10248.
* #12740.
* #13039.
* #16212.

### CODE REVIEW

File: src/netaddress.h [hunk 1]
```diff
@@ -161,7 +161,7 @@ class CService : public CNetAddr
         inline void SerializationOp(Stream& s, Operation ser_action) {
             READWRITE(ip);
             unsigned short portN = htons(port);
-            READWRITE(FLATDATA(portN));
+            READWRITE(portN);
```
Comment (random-zebra) [inline src/netaddress.h:164]: Mhm.. I don't think we can simply do this.
This is not a char (or char-array), for which we introduced the native serialization. This is a short int, stored in memory in native byte order, but serialized in big endian notation (see bitcoin#12886 and bitcoin#12916)

**Review by random-zebra — COMMENTED** [review]
Looking good. Just a compatibility concern with the last commit.

File: src/netaddress.h (hunk 1 above)
Comment (furszy) [reply src/netaddress.h:164]: hmm yeah, nice catch.

**Comment by furszy** [comment]
Updated per feedback

**Review by random-zebra — APPROVED** [review]
ACK d593e7eca0fc5794e56a38506a1f27d54f60363a

**Review by Fuzzbawls — APPROVED** [review]
ACK d593e7eca0fc5794e56a38506a1f27d54f60363a

with a minor nit

File: src/netaddress.h [hunk 2]
```diff
@@ -159,15 +157,14 @@ class CService : public CNetAddr
...
-            port = ntohs(portN);
-    }
+        template <typename Stream, typename Operation>
+        inline void SerializationOp(Stream& s, Operation ser_action) {
+            READWRITE(ip);
+            unsigned short portN = htons(port);
+            READWRITE(FLATDATA(portN));
+            if (ser_action.ForRead())
+                 port = ntohs(portN);
+        }
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/netaddress.h:167]: we should also restore our styling convention of branching on newlines for functions implementations here.
Merging this for now. We can fix the styling in a separate trivial PR.

**Pull request merged**